### PR TITLE
Generate cases for all ScalarTypes in Type functions

### PR DESCRIPTION
**Pull request by li-roy**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #19295 Generate only one Type class per backend
* #19275 Make complex its own backend
* #19270 Add ScalarType argument to Type::options()
* **#19230 Generate cases for all ScalarTypes in Type functions that call to TH**
* #19183 Remove usages of TypeID
* #18972 Make copy_ a native function, remove copy and copy_ from Type

- Changed all Type classes of the same backend to generate the same functions
- Functions now use a switch statement based on the dispatch scalar type
- Example for a TH and native function: https://gist.github.com/li-roy/f74db7130c3e45f69d08dc8903cdb050
- Note that I couldn't use our Dispatch macros because some functions need to return a Tensor &, but our Dispatch macros use lambdas that infer a temporary return type.

Differential Revision: [D14927991](https://our.internmc.facebook.com/intern/diff/D14927991)

### CODE REVIEW

**Comment by li-roy** [comment]
closing #18603 in favor of this.

**Review by smessmer — APPROVED** [review]

File: aten/src/ATen/templates/TypeDerived.cpp [hunk 1]
```diff
@@ -15,6 +15,7 @@
 #include <ATen/NativeFunctions.h>
 #include <ATen/Utils.h>
 #include <ATen/WrapDimUtils.h>
+#include <ATen/Dispatch.h>
```
Comment (gchanan) [inline aten/src/ATen/templates/TypeDerived.cpp]: I don't think you need this.

File: aten/src/ATen/templates/SparseTypeDerived.cpp [hunk 2]
```diff
@@ -13,6 +13,7 @@
 #include <ATen/NativeFunctions.h>
 #include <ATen/Utils.h>
 #include <ATen/WrapDimUtils.h>
+#include <ATen/Dispatch.h>
```
Comment (gchanan) [inline aten/src/ATen/templates/SparseTypeDerived.cpp]: I don't think you need this.

File: aten/src/ATen/core/Type.h [hunk 3]
```diff
@@ -94,7 +94,6 @@ enum class TypeID {
   MSNPULong,
   MSNPUShort,
   MSNPUHalf,
-  MSNPUQInt8,
```
Comment (gchanan) [inline aten/src/ATen/core/Type.h]: why did these go away?  How are backend extensions supposed to implement quantization?

File: aten/src/ATen/gen.py [hunk 4]
```diff
@@ -180,7 +181,7 @@ def backend_to_devicetype(backend):
         return 'CPU'
     return backend
 
-backends = ['CPU', 'CUDA', 'QuantizedCPU']
+backends = ['CPU', 'CUDA']
```
Comment (gchanan) [inline aten/src/ATen/gen.py]: can you clean these up in the future?  These aren't backends, they are device types.

File: aten/src/ATen/preprocess_declarations.py [hunk 5]
```diff
@@ -31,46 +33,60 @@ def process_types_and_backends(option):
...
 
-        types = option.get('types', all_types)
-
-        pairs = [[p, t] for p in backends for t in types]
+        backend_types = {}
+        for backend in backends:
+            if backend == 'QuantizedCPU':
+                backend_types[backend] = type_map['quantized']
+            else:
+                backend_types[backend] = option.get('types', all_types)
```
Comment (gchanan) [inline aten/src/ATen/preprocess_declarations.py]: wait, QInt8 is included in here?  Even though QInt8 is never included in this "backend"?  What's going on here?

**Pull request closed without merging**